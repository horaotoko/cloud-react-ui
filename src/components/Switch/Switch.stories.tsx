import { Story, Meta } from "@storybook/react";
import { Switch, SwitchProps } from "./Switch";

export default {
  component: Switch,
  title: "components/Switch",
} as Meta;

const Template: Story<SwitchProps> = (args) => <Switch {...args} />;

export const Default = Template.bind({});
