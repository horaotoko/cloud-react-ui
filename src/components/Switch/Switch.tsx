import { forwardRef } from "react";
import MuiSwitch, {
  SwitchProps as MuiSwitchProps,
} from "@material-ui/core/Switch";
import { makeStyles } from "@material-ui/core/styles";
import { color } from "../../theme";

export type SwitchProps = MuiSwitchProps & {
  className?: string;
};

const useStyles = makeStyles(() => ({
  root: {
    width: 32,
    height: 20,
    padding: 0,
  },
  switchBase: {
    color: color.white,
    padding: 2,
    "&$checked": {
      transform: "translateX(12px)",
      color: color.white,
      "& + $track": {
        opacity: 1,
        backgroundColor: color.royalBlue,
        borderColor: color.royalBlue,
      },
    },
  },
  thumb: {
    width: 16,
    height: 16,
    boxShadow: "0 1px 4px rgba(0, 0, 0, .15)",
  },
  track: {
    borderRadius: 20 / 2,
    backgroundColor: color.linkWater,
    opacity: 1,
  },
  checked: {},
}));

export const Switch = forwardRef<HTMLButtonElement, SwitchProps>(
  (props, ref) => {
    const classes = useStyles();
    return (
      <MuiSwitch
        disableRipple
        classes={{
          root: classes.root,
          switchBase: classes.switchBase,
          thumb: classes.thumb,
          track: classes.track,
          checked: classes.checked,
        }}
        {...props}
        ref={ref}
      />
    );
  }
);

Switch.displayName = "Switch";
