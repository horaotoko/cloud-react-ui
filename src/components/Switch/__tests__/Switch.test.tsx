import { render, fireEvent } from '@testing-library/react';
import { Switch } from '../Switch';

describe('Switch', () => {
  it('default', () => {
    const { asFragment, getByRole } = render(<Switch />);
    expect(asFragment()).toMatchSnapshot();

    const checkbox = getByRole('checkbox');
    expect(checkbox).not.toBeChecked();
    expect(checkbox).not.toBeDisabled();
  });

  it('defaultChecked props', () => {
    const { getByRole } = render(<Switch defaultChecked={true} />);
    expect(getByRole('checkbox')).toBeChecked();
  });

  it('disabled props', () => {
    const { getByRole } = render(<Switch disabled={true} />);
    expect(getByRole('checkbox')).toBeDisabled();
  });

  it('onChange, onClick props', () => {
    const onChange = jest.fn();
    const onClick = jest.fn();
    const { getByRole } = render(
      <Switch onChange={onChange} onClick={onClick} />
    );
    fireEvent.click(getByRole('checkbox'));

    expect(onChange).toBeCalledTimes(1);
    expect(onClick).toBeCalledTimes(1);
  });
});
