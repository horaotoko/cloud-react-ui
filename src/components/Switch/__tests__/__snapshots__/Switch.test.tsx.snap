// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch default 1`] = `
<DocumentFragment>
  <span
    class="MuiSwitch-root makeStyles-root-1"
  >
    <span
      aria-disabled="false"
      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiSwitch-switchBase makeStyles-switchBase-2 MuiSwitch-colorSecondary"
    >
      <span
        class="MuiIconButton-label"
      >
        <input
          class="PrivateSwitchBase-input-9 MuiSwitch-input"
          type="checkbox"
          value=""
        />
        <span
          class="MuiSwitch-thumb makeStyles-thumb-3"
        />
      </span>
    </span>
    <span
      class="MuiSwitch-track makeStyles-track-4"
    />
  </span>
</DocumentFragment>
`;
